packages/jest-cli/src/search_source.js

```diff
@@ -183,27 +177,28 @@ class SearchSource {
     return {tests: []};
   }
 
-  async findChangedTests(options: Options): Promise<SearchResult> {
-    const {repos, changedFiles} = await getChangedFilesForRoots(
-      this._context.config.roots,
-      options,
-    );
+  async findTestRelatedToChangedFiles(
+    changedFilesPromise: ChangedFilesPromise,
+  ) {
+    const {repos, changedFiles} = await changedFilesPromise;
 
     // no SCM (git/hg/...) is found in any of the roots.
     const noSCM = Object.keys(repos).every(scm => repos[scm].size === 0);
-
     return noSCM
       ? {noSCM: true, tests: []}
       : this.findRelatedTests(changedFiles);
   }
 
-  getTestPaths(
+  async getTestPaths(
     testSelectionConfig: TestSelectionConfig,
+    changedFilesPromise: ?ChangedFilesPromise,
   ): Promise<SearchResult> {
     if (testSelectionConfig.onlyChanged) {
-      return this.findChangedTests({
-        lastCommit: testSelectionConfig.lastCommit,
-      });
+      if (!changedFilesPromise) {
+        throw new Error('This promise must be present when running with -o.');
```

Comment:
Should we change the error message to something like "'A Git or Mercurial repository must be present when running with -o."? So that it gives some info to the user